osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/ThrottlingService.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2020 Smart Society Services B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.application.services;
+
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ThrottlingService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThrottlingService.class);
+
+    @Value("${throttling.max.open.connections}")
+    private int maxOpenConnections;
+    @Value("${throttling.max.new.connection.requests}")
+    private int maxNewConnectionRequests;
+    @Value("${throttling.reset.time}")
+    private int resetTime;
+
+    private Semaphore openConnectionsSemaphore;
+    private Semaphore newConnectionRequestsSemaphore;
+    private Timer resetTimer;
+    private ReentrantLock resetTimerLock;
+
+    @PostConstruct
+    public void postConstruct() {
+
+        openConnectionsSemaphore = new Semaphore(maxOpenConnections);
+        newConnectionRequestsSemaphore = new Semaphore(maxNewConnectionRequests);
+
+        resetTimerLock = new ReentrantLock();
+
+        resetTimer = new Timer();
+        resetTimer.scheduleAtFixedRate(new ResetTask(), resetTime, resetTime);
+
+        LOGGER.info("Initialized ThrottlingService. {}", this);
+    }
+
+    @PreDestroy
+    public void preDestroy() {
+        if (this.resetTimer != null) {
+            this.resetTimer.cancel();
+        }
+    }
+
+    public void openConnection() {
+
+        LOGGER.debug("Requesting openConnection. available = {} ", this.openConnectionsSemaphore.availablePermits());
+
+        try {
+            this.openConnectionsSemaphore.acquire();
+            LOGGER.debug("openConnection granted. available = {} ", this.openConnectionsSemaphore.availablePermits());
+        } catch (InterruptedException e) {
+            LOGGER.warn("Unable to acquire Open Connection", e);
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public void closeConnection() {
+
+        LOGGER.debug("closeConnection(). available = {} ", this.openConnectionsSemaphore.availablePermits());
+        this.openConnectionsSemaphore.release();
+    }
+
+    public void newConnectionRequest() {
```

Comment:
Ik verwacht dat newConnectionRequest vooraf gaat aan openConnection. M.a.w. je mag geen openConnection aanroepen voordat newConnectionRequest goed gegaan is. Is het een idee om de call naar newConnectionRequest te doen in openConnection()? Hiermee zou je voorkomen dat de verantwoordelijkheid om beide aan te roepen niet bij de client ligt en dus foutief geimplementeerd kan worden.